engine/src/Interpreter/interpreter_ops.cuh

```diff
@@ -152,512 +114,264 @@ public:
+			write_valid_data(column_index, valids_out_buffer[column_index], row_index);
 		}
 	}
 
-	__device__ __forceinline__ void copy_row_valids_Into_global(
-		int64_t * valids_out_buffer, const cudf::size_type & row_index) {
-		for(column_index_type out_index = 0; out_index < this->num_final_outputs; out_index++) {
-			if(this->valid_ptrs_out[out_index] != nullptr) {
-				write_valid_data(out_index, valids_out_buffer[out_index], row_index);
-			}
-		}
-	}
-
-	__device__ __forceinline__ void load_cur_row_valids(
-		int64_t valids_in_buffer[], cudf::size_type row, int64_t & cur_row_valids, column_index_type num_columns) {
-		for(column_index_type cur_column = 0; cur_column < num_columns; cur_column++) {
-			setColumnValid(cur_row_valids, cur_column, getColumnValid(valids_in_buffer[cur_column], row));
-		}
-	}
+private:
 
 	/**
 	 * @param buffer the local buffer which storse the information that is to be processed
 	 * @param position the position in the local buffer where this data needs to be written
 	 */
-	template <typename LocalStorageType>
-	__device__ __forceinline__ LocalStorageType get_data_from_buffer(int64_t * buffer, int position) {
-		return *((LocalStorageType *) (buffer + ((position * threadBlockSize) + threadIdx.x)));
+	CUDA_DEVICE_CALLABLE void get_data_from_buffer(int64_t * data, int64_t * buffer, int position) {
+		int64_t * register_ptr = buffer + (position * blockDim.x + threadIdx.x);
+		*data = *register_ptr;
+	}
+
+	CUDA_DEVICE_CALLABLE void get_data_from_buffer(double * data, int64_t * buffer, int position) {
+		int64_t * register_ptr = buffer + (position * blockDim.x + threadIdx.x);
+		*data = __longlong_as_double(*register_ptr);
 	}
 
-	template <typename LocalStorageType>
-	__device__ __forceinline__ void store_data_in_buffer(LocalStorageType data, int64_t * buffer, int position) {
-		*((LocalStorageType *) (buffer + ((position * threadBlockSize) + threadIdx.x))) = data;
+	CUDA_DEVICE_CALLABLE void store_data_in_buffer(int64_t data, int64_t * buffer, int position) {
+		int64_t * register_ptr = buffer + (position * blockDim.x + threadIdx.x);
+		*register_ptr = data;
+	}
+
+	CUDA_DEVICE_CALLABLE void store_data_in_buffer(double data, int64_t * buffer, int position) {
+		int64_t * register_ptr = buffer + (position * blockDim.x + threadIdx.x);
+		*register_ptr = __double_as_longlong(data);
 	}
 
 	/**
 	 * @param buffer the local buffer which storse the information that is to be processed
 	 */
-	template <typename ColType, typename LocalStorageType>
-	__device__ __forceinline__ void device_ptr_read_into_buffer(column_index_type col_index,
-		cudf::size_type row,
-		const void * const * columns,
-		int64_t * buffer,
-		int position) {
-		const ColType * col_data = static_cast<const ColType *>((columns[col_index]));
-		*((LocalStorageType *) (buffer + ((position * threadBlockSize) + threadIdx.x))) =
-			(LocalStorageType) __ldg(((ColType *) &col_data[row]));
+	struct device_ptr_read_into_buffer {	
```

Comment:
I prefer the template where there is a typename for the LocalStorageType it makes it much clearer what is happening

Reply:
I need to specialize for the different data types, the buffer is always `int64_t *` so no need for another tempate argument, just need to apply the correct conversion like `__double_as_longlong`